src/topology/bornology/basic.lean

```diff
@@ -0,0 +1,137 @@
+
+variables {α β : Type*}
+
+
+/-- A **bornology** on a type `α` is a filter of cobounded sets which contains the cofinite filter.
+Such spaces are equivalently specified by their bounded sets, see `bornology.of_bounded`
+and `bornology.eq_iff_is_bounded`-/
+@[ext]
+class bornology (α : Type*) :=
+(cobounded [] : filter α)
+(le_cofinite [] : cobounded ≤ cofinite)
+
+attribute [class] bornology
+
+/-- A constructor for bornologies by specifying the bounded sets,
+and showing that they satisfy the appropriate conditions. -/
+def bornology.of_bounded {α : Type*} (B : set (set α))
+  (empty_mem : ∅ ∈ B) (subset_mem : ∀ s₁ ∈ B, ∀ s₂ : set α, s₂ ⊆ s₁ → s₂ ∈ B)
+  (union_mem : ∀ s₁ s₂ ∈ B, s₁ ∪ s₂ ∈ B) (sUnion_univ : ⋃₀ B = univ) :
+  bornology α :=
+{ cobounded :=
+  { sets := {s : set α | sᶜ ∈ B},
+    univ_sets := by { rwa ←compl_univ at empty_mem },
+    sets_of_superset := λ x y hx hy, subset_mem xᶜ hx yᶜ (compl_subset_compl.mpr hy),
+    inter_sets := λ x y hx hy, by simpa [compl_inter] using union_mem xᶜ hx yᶜ hy, },
+  le_cofinite :=
+  begin
+    refine le_def.mpr (λ s, _),
+    simp only [mem_set_of_eq, mem_cofinite, filter.mem_mk],
+    generalize : sᶜ = s',
+    refine λ h, h.dinduction_on _ (λ x t hx ht h, _),
+    { exact empty_mem, },
+    { refine insert_eq x t ▸ union_mem _ _ _ h,
+      obtain ⟨b, hb : b ∈ B, hxb : x ∈ b⟩ :=
+        mem_sUnion.mp (by simpa [←sUnion_univ] using mem_univ x),
+      exact subset_mem _ hb _ (singleton_subset_iff.mpr hxb) },
+  end }
+
+namespace bornology
+
+section
+variables [bornology α] {s₁ s₂ : set α}
+
+/-- `is_cobounded` is the predicate that `s` is in the filter of cobounded sets in the ambient
+bornology on `α` -/
+def is_cobounded (s : set α) : Prop := s ∈ cobounded α
+
+/-- `is_bounded` is the predicate that `s` is bounded relative to the ambient bornology on `α`. -/
+def is_bounded (s : set α) : Prop := sᶜ ∈ cobounded α
+
+lemma is_cobounded_def {s : set α} : is_cobounded s ↔ s ∈ cobounded α := iff.rfl
+
+lemma is_bounded_def {s : set α} : is_bounded s ↔ sᶜ ∈ cobounded α := iff.rfl
+
+lemma is_bounded_compl_iff {s : set α} : is_bounded sᶜ ↔ is_cobounded s :=
+by rw [is_bounded_def, is_cobounded_def, compl_compl]
+
+@[simp]
+lemma is_bounded_empty : is_bounded (∅ : set α) :=
+by { rw [is_bounded_def, compl_empty], exact univ_mem}
+
+lemma is_bounded.union (h₁ : is_bounded s₁) (h₂ : is_bounded s₂) : is_bounded (s₁ ∪ s₂) :=
+by { rw [is_bounded_def, compl_union], exact (cobounded α).inter_sets h₁ h₂ }
+
+lemma is_bounded.subset (h₁ : is_bounded s₂) (h₂ : s₁ ⊆ s₂) : is_bounded s₁ :=
+by { rw [is_bounded_def], exact (cobounded α).sets_of_superset h₁ (compl_subset_compl.mpr h₂) }
+
+@[simp]
+lemma sUnion_bounded_univ : (⋃₀ {s : set α | is_bounded s}) = set.univ :=
+univ_subset_iff.mp $ λ x hx, mem_sUnion_of_mem (mem_singleton x)
+  $ le_def.mp (le_cofinite α) {x}ᶜ $ (set.finite_singleton x).compl_mem_cofinite
+
+end
+
+end bornology
+
+namespace bornology
+
+lemma eq_iff {t t' : bornology α} :
```

Comment:
should this maybe be called `ext_iff`? and also provide an `@[ext] lemma ext` version for the tactic?

Reply:
or are you thinking maybe the one below may be wanting to be used a similar amount of times? anyways, I do still think they should say `ext` somewhere

Reply:
yes, I'm not sure which will be used more frequently, and also the class itself has the `@[ext]` attribute, so I can't call it `ext_iff`. I have changed the names to `ext_iff'` and `ext_iff_of_bounded` though.